### PR TITLE
update docs

### PR DESCRIPTION
**Pull request by nickeubank**
Input welcome!

### CODE REVIEW

File: docs/src/basicproperties.md [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+- `neighbors`: Return array of neighbors of a vertex. If graph is directed, output is equivalent of `out_neighbors`.
+- `all_neighbors`:  Returns array of all neighbors (both `in_neighbors` and `out_neighbors`). For undirected graphs, equivalent to `neighbors`.
+- `in_neighbors`: Return array of in-neighbors. Equivalent to `neighbors` for undirected graphs.
+- `out_neighbors`: Return array of out-neighbors. Equivalent to `neighbors` for undirected graphs.
+
+## Edge Properties
+
+- `src`: Give source vertex of an edge.
+- `dst`: Give destination vertex of an edge.
+- `reverse`: Creates a new edge running in opposite direction of passed edge.
```
Comment (sbromberger) [inline docs/src/basicproperties.md]: I think we should just reference the functions here and let the docstrings provide the documentation. That way, if we ever change functionality, we won't have to remember updating multiple places.

File: docs/src/generators.md [hunk 2]
```diff
@@ -1,49 +1,51 @@
...
-Other classical graphs can be generated by the following function:
+## Datasets
 
-```@autodocs
-Modules = [LightGraphs]
-Pages   = [ "generators/smallgraphs.jl"]
-Private = false
-```
+Other notorious graphs and integration with the `MatrixDepot.jl` package are available in the `Datasets` submodule of the companion package
+*LightGraphsExtras.jl*
```
Comment (sbromberger) [inline docs/src/generators.md]: This will eventually move out into its own package.

File: docs/src/index.md [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+julia> g = PathGraph(6)
+
+# Number vertices
+julia> nv(g)
+
+# Number edges
+julia> ne(g)
+
+# Add an edge to make the path a loop
+julia> add_edge!(g, [6, 1])
```
Comment (sbromberger) [inline docs/src/index.md]: ergh. I don't think we should be highlighting edge addition using a vector. `add_edge!(g, e)` or `add_edge!(g, s, d)` are the two preferred ways of doing it.

File: docs/src/persistence.md [hunk 4]
```diff
@@ -1,14 +1,10 @@
...
 
-Graphs may be written to I/O streams and files using the `savegraph` function and read with the `loadgraph` function. The default graph format is a proprietary compressed *LightGraphs.jl* format `lg`. Other formats are available via the *GraphPersistence* package.
+## Saving using *LightGraphs.jl* `lg` format.
 
-```@autodocs
-Modules = [LightGraphs]
-Pages   = [ "persistence/common.jl"]
-Private = false
-```
+Graphs may be written to I/O streams and files using the `savegraph` function and read with the `loadgraph` function. The default graph format is a proprietary compressed *LightGraphs.jl* format `lg`.
```
Comment (sbromberger) [inline docs/src/persistence.md]: Note that other graph implementations add methods to `savegraph` and `loadgraph`, so these functions are not limited to SimpleGraphs.

File: docs/src/types.md [hunk 5]
```diff
@@ -0,0 +1,44 @@
+#LightGraph Types
+
+The LightGraph library supports both the `AbstractGraph` type and two concrete simple graph types- - `Graph` for undirected graphs and `DiGraph` for directed graphs -- that are subtypes of `AbstractGraph`.
+
+## SimpleGraph Types
+
+*LightGraphs.jl* provides two concrete graph types (referred to as "simple graphs"): `Graph` is an undirected graph, and `DiGraph` is its directed counterpart. Both of these types can be parameterized to specifying how vertices are identified (by default, `Graph` and `DiGraph` use the system default integer type, usually `Int64`).
```
Comment (sbromberger) [inline docs/src/types.md]: Technically, `Graph` and `DiGraph` are aliases for `SimpleGraph` and `SimpleDiGraph` used to provide backwards compatibility.

Comment (nickeubank) [reply docs/src/types.md]: @sbromberger this has definitely always confused me a little. Is the best to just say `Graph` and `DiGraph` and not bring the "simple" component into it?

Comment (sbromberger) [reply docs/src/types.md]: Yeah, it's confusing to me too. At some point what I'd like to do is get rid of `Graph` and `DiGraph` as type aliases and maybe do what Plots does with its backend: that is, load a graph type (or types) and then let the user set a default.

File: README.md [hunk 6]
```diff
@@ -27,118 +25,28 @@ other information, for example) is expected to be stored outside of the graph
 structure itself. Such data lends itself to storage in more traditional and
 better-optimized mechanisms.
 
-Additional functionality may be found in the companion package [LightGraphsExtras.jl](https://github.com/JuliaGraphs/LightGraphsExtras.jl).
+Additional functionality may be found in a number of companion packages, including:
```
Comment (jpfairbanks) [inline README.md]: Maybe this should say something like:

> LightGraphs.jl is the central package of the JuliaGraphs ecosystem. Additional functionality such as, advanced IO and file formats, weighted graphs, property graphs, and optimization related functions can be found in the following packages.

And then sort the list in this order. I think this order is most likely to be clicked to least likely.

File: docs/src/index.md (hunk 3 above)
Comment (jpfairbanks) [reply docs/src/index.md]: Good, that would be SLOW!

File: docs/src/persistence.md (hunk 4 above)
Comment (jpfairbanks) [reply docs/src/persistence.md]: Proprietary is not really correct in this usage. The format is FOSS licensed just like the software that defines it. We mean `custom` or `bespoke` not `commercial` or `copy-righted`

Definition of proprietary from Merriam Webster:

> 2: something that is used, produced, or marketed under exclusive legal right of the inventor or maker; specifically :a drug (such as a patent medicine) that is protected by secrecy, patent, or copyright against free competition as to name, product, composition, or process of manufacture

File: docs/src/types.md (hunk 5 above)
Comment (jpfairbanks) [reply docs/src/types.md]: Yeah that is a little tricky, I am in favor of encouraging people to use `Graph/DiGraph` names and then if they want to change to a different concrete implementation they can do

```julia
using LightGraphs
using SimpleWeightedGraphs
import SimpleWeightedGraphs: SimpleWeightedGraph

Graph = SimpleWeightedGraph

g = Graph(n)
...
```

File: docs/src/types.md [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+
+A graph *G* is described by a set of vertices *V* and edges *E*: *G = {V, E}*. *V* is an integer range `1:n`; *E* is represented as forward (and, for directed graphs, backward) adjacency lists indexed by vertices. Edges may also be accessed via an iterator that yields `Edge` types containing `(src<:Integer, dst<:Integer)` values. Both vertices and edges may be integers of any type, and the smallest type that fits the data is recommended in order to save memory.
+
+Graphs are created using `Graph()` or `DiGraph()`; there are several options (see the tutorials for examples).
+
+Multiple edges between two given vertices are not allowed: an attempt to add an edge that already exists in a graph will result in a silent failure.
```
Comment (jpfairbanks) [inline docs/src/types.md]: > Multiple edges between two given vertices are not allowed: an attempt to add an edge that already exists in a graph will not raise an error. This event can be detected using the return value of `add_edge!`.

File: docs/src/types.md [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+
+
+## AbstractGraph Type
+
+To facilitate the development of other concrete graph types, *LightGraphs.jl* implements the `AbstractGraph` type, which is used by libraries like [MetaGraphs.jl](https://github.com/JuliaGraphs/MetaGraphs.jl) (for graphs with associated meta-data) and [SimpleWeightedGraphs.jl](https://github.com/JuliaGraphs/SimpleWeightedGraphs.jl) (for weighted graphs). All types that are a subset of `AbstractGraph` must implement the following functions (most of which are described in more detail in [Accessing Graph Properties](@ref) and [Making and Modifying Graphs](@ref)):
```
Comment (jpfairbanks) [inline docs/src/types.md]: > To encourage experimentation and development within the JuliaGraphs ecosystem, *LightGraphs.jl* defines the `AbstractGraph` type, which...

File: docs/src/index.md [hunk 9]
```diff
@@ -0,0 +1,31 @@
+# Light Graphs
+
+The goal of *LightGraphs.jl* is to mirror the functionality of robust network and graph analysis libraries such as [NetworkX](http://networkx.github.io) while being simpler to use and more efficient than existing Julian graph libraries such as [Graphs.jl](https://github.com/JuliaLang/Graphs.jl). It is an explicit design decision that any data not required for graph manipulation (attributes and other information, for example) is expected to be stored outside of the graph structure itself. Such data lends itself to storage in more traditional and better-optimized mechanisms.
```
Comment (jpfairbanks) [inline docs/src/index.md]: This is a foundational manifesto that we have shed with maturity. We are no longer defined by our relationship to NetworkX and Graphs.jl. We have our own fate in our own hands! Long Live LightGraphs!

File: docs/src/generators.md [hunk 10]
```diff
@@ -1,49 +1,51 @@
...
+- `add_edge!`: Add an edge to an existing graph.
+- `rem_edge!`: Remove an edge from an existing graph.
+- `add_vertex!`, `add_vertices!`: Add one (or several) vertices to an existing graph.
+- `rem_vertex!`: Remove an existing vertex. Note that for `Graph` and `DiGraph` objects
+this is accomplished via a "swap-and-pop" procedure where the highest-indexed vertex
+is swapped with the vertex to be removed, and only then is the vertex removed. As a
+result, unlike in some libraries (like `iGraph`), vertex indices will not all
+down-shift when a vertex is removed; rather, the only change is that the previously
+highest-index vertex will take on the index of the removed index.
+- `zero`: creates a zero-vertex zero-edge graph of same types as the graph passed
```
Comment (jpfairbanks) [inline docs/src/generators.md]: creates an empty graph, with no vertices and no edges graph with the same types as the input graph.

File: docs/src/basicproperties.md [hunk 11]
```diff
@@ -0,0 +1,27 @@
+# Accessing Graph Properties
+The following is an overview of functions for accessing graph properties. For functions that modify graphs, see [Making and Modifying Graphs](@ref). 
+
+## Whole-Graph Properties:
```
Comment (jpfairbanks) [inline docs/src/basicproperties.md]: "Graph Properties"? Whole Graph sounds like Whole Wheat to me.

File: docs/src/persistence.md (hunk 4 above)
Comment (sbromberger) [reply docs/src/persistence.md]: +1 to "bespoke". Though - @jpfairbanks - it'd be great to have an mmaped file format :)

File: docs/src/generators.md (hunk 10 above)
Comment (sbromberger) [reply docs/src/generators.md]: Again, I think we should let the docstrings speak for themselves here.

File: docs/src/basicproperties.md (hunk 11 above)
Comment (sbromberger) [reply docs/src/basicproperties.md]: `struct BranGraph <: AbstractGraph`

File: docs/src/basicproperties.md (hunk 1 above)
Comment (nickeubank) [reply docs/src/basicproperties.md]: I think there's something to be said for a short, clean, concise presentation of the core methods. The autodocs are comprehensive, but they have a lot of visual noise and can be overwhelming -- basically, I think at least one stream-lined presentation is really helpful for potential new users.

File: docs/src/index.md (hunk 3 above)
Comment (nickeubank) [reply docs/src/index.md]: oops! Ok!

Comment (nickeubank) [reply docs/src/index.md]: Is there a way to actually get the doc builder to run blocks of code (like with `sphinx` for python libraries) so those things get caught more easily?

File: docs/src/basicproperties.md (hunk 1 above)
Comment (nickeubank) [reply docs/src/basicproperties.md]: (But I will remove the other places I copy-pasted like in defining the API for AbstractGraphs)

File: docs/src/persistence.md (hunk 4 above)
Comment (nickeubank) [reply docs/src/persistence.md]: hahaha -- that was copy paste off the old docs! Updating to bespoke now.

File: README.md (hunk 6 above)
Comment (nickeubank) [reply README.md]: ✓

**Comment by nickeubank** [comment]
OK, comments integrated. Also broke out plotting into its own page. 

I did not convert the ``Accessing Properties`` into an autodoc -- as noted inline, I think it's really important for new users to be able to see a short summary of core functions in a clean, concise manner, and I can't find a way to do it with auto-docs, though I'm open to suggestions of how to do it better.

**Comment by sbromberger** [comment]
Sorry I've been incommunicado – I've been out of town. I promise to look at this by early next week. Thanks for sticking with it.

**Comment by sbromberger** [comment]
I think this looks good. I'll merge it as soon as my minor edits pass tests.

File: docs/src/persistence.md [hunk 12]
```diff
@@ -1,14 +1,10 @@
...
-Pages   = [ "persistence/common.jl"]
-Private = false
-```
+Graphs may be written to I/O streams and files using the `savegraph` function and read with the `loadgraph` function. The default graph format is a bespoke compressed *LightGraphs.jl* format `lg`.
 
-## Examples
+### Example
 
 ```julia
 save(STDOUT, g)
```
Comment (nickeubank) [inline docs/src/persistence.md]: @sbromberger It occurs to me this example doesn't quite jive with the text. What is the preferred way to save / load now?

Comment (sbromberger) [reply docs/src/persistence.md]: `savegraph` / `savegraphs`

**Comment by nickeubank** [comment]
@sbromberger  ok, updated saving examples and text in last commit.

**Comment by sbromberger** [comment]
@nickeubank -sorry that took so long. Just got back into town.

**Comment by nickeubank** [comment]
@sbromberger no worries, thanks!

**Pull request merged**